### PR TITLE
Scripts/Ulduar: Fix Stormcaller Brundir walking into walls

### LINKED ISSUE #23751: Stormcaller Brundir enters into walls / unkillable
**Description:**

A lot of times Stormcaller Brundir ( 32857 )enters into walls or stops on top of some places making him to recover hp

**Current behaviour:**

It fully breaks the event.

**Expected behaviour:**

It never must enter walls/ stop on unreachable places making him to recover hp.

**Steps to reproduce the problem:**

1. go ulduar
2. fight brundir killing the other 2 bosses,
3. run around to force him to fly.

https://i.imgur.com/FDdOu63.jpg

**Branch(es):**

3.3.5a, for sure all

**TC rev. hash/commit:** 

 4074cb0f01d1 for sure all since implementation, i got reports years ago.

**Operating system:** Debian 10 x64, Ubuntu 19.10 x64

### PR DESCRIPTION
**Pull request by Artamedes**
**Changes proposed:**

-  Change logic of moving, so it doesn't move out of room to begin with

**Target branch(es):** 3.3.5/master

- [x] 3.3.5

**Issues addressed:** Closes #23751

**Tests performed:** (Does it build, tested in-game, etc.)
Builds and tested ingame

**Known issues and TODO list:** (add/remove lines as needed)

- [x] test!

### CODE REVIEW

**Comment by illfated** [comment]
Hope it works (I have not tested it yet, maybe Aokromes wants to test it).
I noticed that there are no blank lines in the rest of the switch section.
Just in case it gets accepted and it works, maybe use the same codestyle.

**Comment by Aokromes** [comment]
@illfated i got a ginea pig to help me and with that i reproduced the bug xd

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp [hunk 1]
```diff
@@ -626,10 +626,16 @@ class boss_stormcaller_brundir : public CreatureScript
                             {
                                 float x = float(irand(-25, 25));
                                 float y = float(irand(-25, 25));
-                                me->GetMotionMaster()->MovePoint(0, me->GetPositionX() + x, me->GetPositionY() + y, me->GetPositionZ());
+
+                                Position pos = *me;
```
Comment (jackpoz) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp]: please use me->GetPosition(), it makes the intent of the code clearer

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp [hunk 2]
```diff
@@ -626,10 +626,16 @@ class boss_stormcaller_brundir : public CreatureScript
...
+
+                                pos.m_positionX += x;
+                                pos.m_positionY += y;
+
+                                me->GetMotionMaster()->MovePoint(0, pos);
                                 // Prevention to go outside the room or into the walls
                                 if (Creature* trigger = me->FindNearestCreature(NPC_WORLD_TRIGGER, 100.0f, true))
-                                    if (me->GetDistance(trigger) >= 50.0f)
+                                    if (pos.GetExactDist2d(trigger) >= 50.0f)
```
Comment (jackpoz) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp]: could you describe this change more in detail ?

**Review by jackpoz — CHANGES_REQUESTED** [review]

**Comment by jackpoz** [comment]
> **Tests performed:** (Does it build, tested in-game, etc.)
> Not tested ingame

please debug your code so no strange issues are added. anyone who will test it will most likely not be a developer

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp (hunk 2 above)
Comment (Artamedes) [reply src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp]: if the calculated Stormcaller Brundir position is more than 50 yards away, move back to the trigger

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp [hunk 3]
```diff
@@ -626,11 +629,20 @@ class boss_stormcaller_brundir : public CreatureScript
...
+
+                                Position pos = me->GetPosition();
+
+                                pos.m_positionX += x;
+                                pos.m_positionY += y;
+
                                 // Prevention to go outside the room or into the walls
-                                if (Creature* trigger = me->FindNearestCreature(NPC_WORLD_TRIGGER, 100.0f, true))
-                                    if (me->GetDistance(trigger) >= 50.0f)
+                                if (auto trigger = ObjectAccessor::GetCreature(*me, m_TriggerGUID))
```
Comment (Shauren) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp]: Do not use `auto` for anything except container iterators (or other super long deep nested types)

File: src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp [hunk 4]
```diff
@@ -495,6 +495,9 @@ class boss_stormcaller_brundir : public CreatureScript
                 events.ScheduleEvent(EVENT_BERSERK, 15min);
                 events.ScheduleEvent(EVENT_CHAIN_LIGHTNING, 4s);
                 events.ScheduleEvent(EVENT_OVERLOAD, urand(60000, 120000));
+
+                if (auto trigger = me->FindNearestCreature(NPC_WORLD_TRIGGER, 100.0f))
```
Comment (Shauren) [inline src/server/scripts/Northrend/Ulduar/Ulduar/boss_assembly_of_iron.cpp]: Don't auto

**Review by Shauren — CHANGES_REQUESTED** [review]

**Comment by Aokromes** [comment]
patch ready on http://tcubuntu.northeurope.cloudapp.azure.com/ to test looking for 2-3 ginea pigs

**Comment by jackpoz** [comment]
> patch ready on http://tcubuntu.northeurope.cloudapp.azure.com/ to test looking for 2-3 ginea pigs

make sure to select the "Test" realm though

**Comment by Aokromes** [comment]
yeah we tested if much better than without patch.

**Comment by jackpoz** [comment]
Thanks for the PR :)

**Pull request merged**